lib/Dialect/FIRRTL/Transforms/LowerBundleVectorTypes.cpp

```diff
@@ -0,0 +1,967 @@
+  auto saoType = sao.input().getType().cast<FVectorType>();
+  auto selectWidth =
+      sao.index().getType().cast<FIRRTLType>().getBitWidthOrSentinel();
+
+  for (size_t index = 0, e = saoType.getNumElements(); index < e; ++index) {
+    auto cond = builder->create<EQPrimOp>(
+        sao.index(),
+        builder->createOrFold<ConstantOp>(UIntType::get(context, selectWidth),
+                                          APInt(selectWidth, index)));
+    builder->create<WhenOp>(cond, false, [&]() {
+      // Recreate the write Path
+      Value leaf = builder->create<SubindexOp>(sao.input(), index);
+      for (int i = writePath.size() - 2; i >= 0; --i)
+        leaf = cloneAccess(builder, writePath[i], leaf);
+
+      if (isa<ConnectOp>(op))
+        builder->create<ConnectOp>(leaf, writePath[0]->getResult(0));
+      else
+        builder->create<PartialConnectOp>(leaf, writePath[0]->getResult(0));
+    });
+  }
+}
+
+// Expand connects of aggregates
+void TypeLoweringVisitor::visitStmt(ConnectOp op) {
+  if (processSAPath(op))
+    return;
+
+  // Attempt to get the bundle types.
+  SmallVector<FlatBundleFieldEntry> fields = peelType(op.dest().getType());
+  if (fields.empty())
+    return;
+
+  // Loop over the leaf aggregates.
+  for (auto field : llvm::enumerate(fields)) {
+    Value src = getSubWhatever(op.src(), field.index());
+    Value dest = getSubWhatever(op.dest(), field.index());
+    if (field.value().isOutput)
+      std::swap(src, dest);
+    builder->create<ConnectOp>(dest, src);
+  }
+  opsToRemove.push_back(op);
+}
+
+void TypeLoweringVisitor::visitStmt(PartialConnectOp op) {
+  if (processSAPath(op))
+    return;
+
+  SmallVector<FlatBundleFieldEntry> srcFields = peelType(op.src().getType());
+  SmallVector<FlatBundleFieldEntry> destFields = peelType(op.dest().getType());
+
+  // Ground Type
+  if (destFields.empty()) {
+    // check for truncation
+    Value src = op.src();
+    Value dest = op.dest();
+    auto srcType = src.getType().cast<FIRRTLType>();
+    auto destType = dest.getType().cast<FIRRTLType>();
+    auto srcWidth = srcType.getBitWidthOrSentinel();
+    auto destWidth = destType.getBitWidthOrSentinel();
+
+    if (destType == srcType) {
+      builder->create<ConnectOp>(dest, src);
+      opsToRemove.push_back(op);
+    } else if (destType.isa<IntType>() && srcType.isa<IntType>() &&
+               destWidth >= 0 && destWidth < srcWidth) {
+      // firrtl.tail always returns uint even for sint operands.
+      IntType tmpType = destType.cast<IntType>();
+      if (tmpType.isSigned())
+        tmpType = UIntType::get(destType.getContext(), destWidth);
+      src = builder->create<TailPrimOp>(tmpType, src, srcWidth - destWidth);
+      // Insert the cast back to signed if needed.
+      if (tmpType != destType)
+        src = builder->create<AsSIntPrimOp>(destType, src);
+      builder->create<ConnectOp>(dest, src);
+      opsToRemove.push_back(op);
+    }
+    return;
```

Comment:
Can the this ever fall through without hitting either the `if` or `else if` branches? This can be simplified a bit with a return in the `if` block if so.